### ISSUE TITLE
vSphere SDK API Endpoint becomes unresponsive - govc: Post https://builds-14.vmware.run/sdk: net/http: TLS handshake timeout

### ISSUE THREAD

**Issue by mhagen-vmware** [vmware/vic#5621]
Seeing this error:
```
09:33:31.081	INFO	Running command 'curl -sk https://192.168.31.114:2378/container-logs.tar.gz -b /tmp/cookies-VCH-12000-3110 | (cd /tmp; tar xvzf - inspiring_mccarthy-bb906f539ebd/tether.debug inspiring_mccarthy-bb906f539ebd/vmware.log) 2>&1'.	
09:33:31.116	INFO	${rc} = 2	
09:33:31.116	INFO	${output} = 
gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

And in vic-admin.log:
```
Jun 29 2017 14:33:38.004Z INFO  User with a valid auth cookie at 192.168.31.248 is authenticated.
Jun 29 2017 14:33:38.007Z DEBUG [BEGIN] [main.(*server).bundleContainerLogs:451]
Jun 29 2017 14:33:38.007Z ERROR Failed to acquire logrotate lock: open /var/run/lock/logrotate_run.lock: permission denied
Jun 29 2017 14:33:38.007Z DEBUG [BEGIN] [main.configureReaders:135]
```

**Comment by mhagen-vmware** [comment]
[Test-Cases.Group9-VIC-Admin.9-01-VICAdmin-ShowHTML-VCH-12000-3110-container-logs.zip](https://github.com/vmware/vic/files/1112565/Test-Cases.Group9-VIC-Admin.9-01-VICAdmin-ShowHTML-VCH-12000-3110-container-logs.zip)

**Comment by sgairo** [comment]
`Jun 29 2017 14:33:23.123Z ERROR Error creating new govmomi client without credentials but with auth cookie: Post https://192.168.31.28/sdk: net/http: TLS handshake timeout
Jun 29 2017 14:33:23.123Z WARN  192.168.31.248:55808 failed to authenticate 
`

**Comment by mhagen-vmware** [comment]
Yep, I think my original description is actually a red herring as I see that on successful runs as well. It looks like perhaps the ESX server just stop responding.  I think we need to add a retry around this call. I have added one in my test now, but we should not bail immediately when the ESX is unresponsive (which can happen easily)

**Comment by mhagen-vmware** [comment]
From @sgairo :
https://github.com/vmware/vic/blob/442ae62449449eb62e3fce18e46b404c0b15a03f/cmd/vicadmin/vicadm.go#L499

**Comment by mhagen-vmware** [comment]
Another instance of the error in a different test:
[Test-Cases.Group9-VIC-Admin.9-02-VICAdmin-CertAuth-VCH-12004-6767-container-logs.zip](https://github.com/vmware/vic/files/1113147/Test-Cases.Group9-VIC-Admin.9-02-VICAdmin-CertAuth-VCH-12004-6767-container-logs.zip)

**Comment by mhagen-vmware** [comment]
In vic-machine also:
```Jun 29 2017 18:25:15.312Z ERROR vic-machine-linux create failed: Failed to connect to 192.168.31.29: Post https://192.168.31.29/sdk: net/http: TLS handshake timeout```

**Comment by mhagen-vmware** [comment]
Manually testing during our CI run:
```
root@4b0da6b89661:/go/src/github.com/vmware/vic# time curl -k https://192.168.31.35/sdk

real	0m0.022s
user	0m0.008s
sys	0m0.000s
root@4b0da6b89661:/go/src/github.com/vmware/vic# time curl -k https://192.168.31.35/sdk

real	0m0.029s
user	0m0.008s
sys	0m0.000s
root@4b0da6b89661:/go/src/github.com/vmware/vic# time curl -k https://192.168.31.35/sdk

real	0m23.820s
user	0m0.012s
sys	0m0.000s
root@4b0da6b89661:/go/src/github.com/vmware/vic# time curl -k https://192.168.31.35/sdk

real	0m0.285s
user	0m0.012s
sys	0m0.004s
root@4b0da6b89661:/go/src/github.com/vmware/vic# time curl -k https://192.168.31.35/sdk

real	0m0.029s
user	0m0.008s
sys	0m0.000s
```

As you can see the time spiked up to 23s in one instance. During this run, the tests did not actually fail, but I think that a spike up to 23s while the average is well less than a second is indicative of a problem - esp when we only have 2 VCHs and a few containers running.

**Comment by mhagen-vmware** [comment]
Seeing a bunch of these in hostd at the time as well:
```
2017-07-06T14:25:55.995Z error hostd[321F2B70] [Originator@6876 sub=SoapAdapter.HTTPService] Failed to read request; stream: <io_obj p:0x382dd62c, h:-1, <TCP '0.0.0.0:0'>, <TCP '0.0.0.0:0'>>, error: N7Vmacore16TimeoutExceptionE(Operation timed out)
2017-07-06T14:26:00.998Z error hostd[324D1B70] [Originator@6876 sub=SoapAdapter.HTTPService] Failed to read request; stream: <io_obj p:0x39fa6a34, h:-1, <TCP '0.0.0.0:0'>, <TCP '0.0.0.0:0'>>, error: N7Vmacore16TimeoutExceptionE(Operation timed out)
2017-07-06T14:26:00.998Z error hostd[32F80B70] [Originator@6876 sub=SoapAdapter.HTTPService] Failed to read request; stream: <io_obj p:0x33094e64, h:-1, <TCP '0.0.0.0:0'>, <TCP '0.0.0.0:0'>>, error: N7Vmacore16TimeoutExceptionE(Operation timed out)
2017-07-06T14:26:00.999Z error hostd[34080B70] [Originator@6876 sub=SoapAdapter.HTTPService] Failed to read request; stream: <io_obj p:0x34ef72dc, h:-1, <TCP '0.0.0.0:0'>, <TCP '0.0.0.0:0'>>, error: N7Vmacore16TimeoutExceptionE(Operation timed out)
```

**Comment by mhagen-vmware** [comment]
[vmsupport-bundle.zip](https://github.com/vmware/vic/files/1128367/vmsupport-bundle.zip)

**Comment by mhagen-vmware** [comment]
Monitoring the task queue doesn't seem to provide any insight, it doesn't seem outrageous...
```
 while sleep 10; do echo `govc host.info -json | jq -r .HostSystems[0].RecentTask | grep Value | wc -l`; done
22
24
24
24
24
27
28
28
28
28
28
29
29
29
29
28
28
31
31
29
33
33
34
34
32
32
32
32
32
30
30
32
33
34
34
34
36
36
36
31
31
34
35
34
36
38
40
40
38
42
42
44
44
38
38
38
41
43
43
43
43
46
49
47
47
51
53
55
50
54
54
56
56
51
55
55
57
55
55
59
59
59
59
59
59
59
57
59
59
59
62
57
57
57
57
57
51
51
54
54
57
51
51
51
52
52
53
53
53
53
54
50
50
56
56
48
48
48
48
43
45
48
48
48
44
44
44
44
36
38
39
39
44
41
govc: Post https://builds-14.vmware.run/sdk: net/http: TLS handshake timeout
0
41
40
42
48
48
42
42
42
43
42
42
44
44
48
41
47
47
47
47
39
42
44
44
44
44
46
46
46
46
40
42
42
42
42
40
40
40
40
42
39
41
41
41
43
36
36
36
36
36
36
36
36
28
28
28
28
28
27
27
27
27
27
21
21
21
23
23
17
17
17
17
17
12
12
12
12
12
10
10
13
13
13
11
13
13
13
13
11
11
13
13
9
9
9
9
11
11
11
11
11
11
13
13
13
13
15
15
15
15
17
17
17
17
19
19
17
17
20
20
20
21
22
22
22
22
22
22
22
22
22
19
19
19
19
19
17
17
17
17
15
15
15
15
15
13
13
13
13
13
13
13
13
13
13
9
9
9
9
9
7
7
7
7
7
5
5
5
5
5
1
1
1
1
1
0
```

**Comment by dougm** [comment]
@mhagen-vmware are you seeing this happen with the master branch or only with PR #5601 ?

How many times have you seen it happen? 

Does it always fail in the same test?

**Comment by dougm** [comment]
If you are able to reproduce, can you also grab the output of these 2 commands:

```
% govc session.ls
% govc host.esxcli network ip connection list
```

**Comment by mhagen-vmware** [comment]
@dougm 
1. currently only in PR #5601, as we have enough delays in our serialized testing to not hit this, but I can reproduce this manually with very little effort and only a single VCH.
2. every time
3. no, random tests in random places. Everywhere we try to hit to the sdk, which is pretty much everywhere in the code.
4. I will repro and grab that output ASAP.

**Comment by dougm** [comment]
Seeing the number of currently poweredOn VCH's get up to 20 or so, each with 2-3 established connections to hostd.  At some point, I'm no longer able to connect to hostd (using govc) and seems hostd crashes at some point.  From what I've seen so far, this seems more like a test issue (too many VCHs running) and/or ESX issue than a VIC issue.

Can you dial down the max number of VCHs that are running at the same time?

**Comment by dougm** [comment]
Keep open a bunch of hostd connections (object.collect with -n is waiting for updates):

```
#!/bin/bash -e

for _ in $(seq 1 150) ; do
  govc object.collect -n=-1 -s - content.about.fullName >/dev/null &
done

wait
```

About ~25 or so result in `govc: Post https://192.168.2.130/sdk: net/http: TLS handshake timeout`

**Comment by mhagen-vmware** [comment]
For this example run, I set parallelization to 4 just for quick repro.  We do see this problem with only 2 VCHs and less than 10 containers.

**Comment by dougm** [comment]
The solution to this will be PR #5784

The idea we discussed of using 1 connection to wait for N tasks doesn't look like its worth the trouble, as the tasks do not take long to complete.

We did end up with a tool to measure task completion times in: vmware/govmomi#806
The method used to watch these tasks could also be used within vic if we find evidence to the contrary regarding the task waiters.

If we are going to look at reducing the number of connections, we should start with WaitForKeyInExtraConfig as this takes much more time than the average task to complete:
https://github.com/vmware/vic/blob/db7a71f4bacf321f99133743ef346898abf5a8e2/pkg/vsphere/vm/vm.go#L184

**Comment by caglar10ur** [comment]
Have we tried increasing the default TLSHandshakeTimeout which is hardcoded as 10 sec. in https://golang.org/src/net/http/transport.go and see what will happen? Looks like we are using the default one (https://github.com/vmware/govmomi/blob/a44f07b26aa46ba0571f5d27a7fb1b69d8ab94f7/vim25/soap/client.go#L115)

IMHO ttat 23s response time clearly shows that ESXi is having bad time.

**Comment by mhagen-vmware** [comment]
We can, that would likely mitigate this issue, but we still need to determine why only 2 VCHs and less than 10 containers VMs is bringing a pretty hefty HW ESXi down like that.

**Comment by mdubya66** [comment]
Needed for parallel testing but not for 1.2.

**Comment by dougm** [comment]
I've been testing against following system:
```console
  Logical CPUs:    4 CPUs @ 3591MHz
  Processor type:  Intel(R) Xeon(R) CPU E5-1620 0 @ 3.60GHz
  Memory:          65461MB
```

3 VCHs x 10 containers, I had issues both with `docker run busybox sleep 5d` and `docker rm -f` when doing all 30 in parallel.  Initially pursed the CPU usage, as both cases have ESX pegged at 100%.  Reducing serial port logging helped a bit (such reverting the debug=1 change via `--debug 0`), but still didn't cut it.

I bumped the `Session.maxInFlight` from 16 to 120 and both the 30 concurrent run + delete tests now pass.  I chose 120 as that was just under the ~125 limit I had seen here: https://github.com/vmware/vic/issues/5621#issuecomment-316849333

The problem was that VIC was timing out in places waiting to get a connection with the limit of 16 while was ESX chugging through the high CPU load.
 
Now with 6.5U1, I don't see hostd issues with that script anymore, 150 open connections is not a problem.  I'll spend more time with this tomorrow and test out Mike's branch too, but at the moment in the short-term it looks like we'll want to:

- Update all 6.5 CI systems to U1

- Increase the connection limit from 16 to X (where X is TBD)

**Comment by mhagen-vmware** [comment]
ah, good idea - I didn't think of the debug 0 idea, not a permanent solution but could help. I spun my build again with that set.  Half of the systems should be on 6.5U1 already, the other half should be on 6.0U3.  Do you think that 6.0U3 is not going to work?

**Comment by dougm** [comment]
I would expect the script to work with 6.0U3, as in 150 concurrent connections should not be a problem.  As for the VIC tests, let's see how we do with 6.5U1 first, did you bump Session.maxInFlight too?

**Comment by mhagen-vmware** [comment]
https://github.com/vmware/vic/pull/5601/files#diff-14cda47afce5189e97493cae2650b242R54

Here is the build:
https://ci.vcna.io/vmware/vic/13454

**Comment by dougm** [comment]
More testing this morning, there's more to do other than bumping maxInFlight, at least on the 'rm' path.  ESX CPU gets pegged with ~1/3 container VMs that haven't fully powered off, finally finishes but takes way longer than it should.  This problem is not related to connections/hostd.

**Comment by hmahmood** [comment]
@dougm what more is required to fix this issue? It sounds like the update to 6.5u1 does not resolve this; is that correct? This is holding up parallelization of test runs per @mhagen-vmware .

**Comment by dougm** [comment]
@hmahmood there were/are multiple issues here.  6.5U1 helped, but was not a silver bullet.  Reverting the debug=1 default helped reduce CPU usage (less serial port logging).  There have been a number of other fixes that have gone in and still in progress related to parallel operations.  In my testing, we do need to bump Session.defaultMaxInFlight higher than 16.. my most recent tests left it at 120 (6.5U1 is required in this case).

As for what else remains, I don't know exactly.  It'll require more debugging & chipping away at the individual issues that turn up.

**Comment by hickeng** [comment]
I've bumped this to high because the connection refused issue was cited as a blocker in getting multiple VCHs working concurrently in VC, and that is necessary for CI improvements.

**Comment by cgtexmex** [comment]
moving back to To Do since #6423 should help -- @mhagen-vmware would be good to see the impact on #5601

**Comment by mhagen-vmware** [comment]
Negative, still hitting this on latest code today:
```
Oct 18 2017 09:39:48.301Z ERROR Error creating new govmomi client without credentials but with auth cookie: Post https://192.168.31.39/sdk: net/http: TLS handshake timeout
```

**Comment by caglar10ur** [comment]
@mhagen-vmware do we have the logs of that run?

**Comment by mhagen-vmware** [comment]
@caglar10ur sure:
https://console.cloud.google.com/m/cloudstorage/b/vic-ci-logs/o/integration_logs_14125_b9a3d068179adc9acbf6749bf20a770067eba23a.zip?authuser=1
10911s
1072

**Comment by caglar10ur** [comment]
@mhagen-vmware thanks!

**Comment by caglar10ur** [comment]
Logs shows that it is the govc binary failing with error and that is expected as VIC changes won't affect govc :) I'm adding a new env. variable to govc to set the timeout

**Comment by caglar10ur** [comment]
https://github.com/vmware/govmomi/pull/890

**Comment by caglar10ur** [comment]
@mhagen-vmware can you update the integration image to use master@govc and try again with exporting GOVC_TLS_HANDSHAKE_TIMEOUT=30s (or passing --tls-handshake-timeout=30s parameter to govc calls) and try again?

**Comment by caglar10ur** [comment]
Any change to get a new run with new govc?

**Comment by mhagen-vmware** [comment]
I tried running it with new govc before, it didn't go well.  Here is another 2 runs:
https://ci.vcna.io/vmware/vic/14344
https://ci.vcna.io/vmware/vic/14345

**Comment by caglar10ur** [comment]
[wrote this to wrong issue so repeating here]

looks like ESXi get restarted while the tests are running

```
Several failures occurred:

1) 'govc: 503 Service Unavailable (Failed to connect to endpoint: [N7Vmacore4Http16LocalServiceSpecE:0x31eb2938] _serverNamespace = /sdk _isRedirect = false _port = 8307)' does not contain 'OK'

2) 'govc: 503 Service Unavailable (Failed to connect to endpoint: [N7Vmacore4Http16LocalServiceSpecE:0x31eb2938] _serverNamespace = /sdk _isRedirect = false _port = 8307)' does not contain 'OK'
```

**Comment by mlh78750** [comment]
Is this also 2005701?

**Comment by hickeng** [comment]
@mlh78750 it certainly sounds like it to me - I'll flag this with needs-info and assign to @mhagen-vmware for verification/refutation now we have root caused 2005701.

**Comment by mhagen-vmware** [comment]
This was the original bug related to not being able to run 2 instances of our tests, not the newer issue related to us leaking VCHs in our new tests.

**Comment by mhagen-vmware** [comment]
It is essentially the same-ish root cause as 2005701 likely, but I think we need a real holistic solution to over running the event queue of the vSphere server.

**Comment by mdubya66** [comment]
lowering to p1. while it is blocking the move to parallel testing that has not yet merged so this is not, yet, a kind/ci-blocker

**Comment by hickeng** [comment]
Closing this issue now it's been broken up into three specific follow up items.

**Comment by renmaosheng** [comment]
similar issue: 
https://ci-vic.vmware.com/vmware/vic/19877/7
from vpxd.log in Test-Cases.Group1-Docker-Commands.1-22-Docker-Volume-RM-VCH-19877-8392-container-logs, at around 2018-08-24T07:38:16.855Z, we can see 

2018-08-24T07:38:16.855Z error vpxd[40935] [Originator@6876 sub=SoapAdapter.HTTPService] Failed to read request; stream: <io_obj p:0x00007fce587f6160, h:-1, <TCP '127.0.0.1 : 8085'>, <TCP '127.0.0.1 : 44828'> FD Closed>, error: N7Vmacore16TimeoutExceptionE(Operation timed out: Stream: <io_obj p:0x00007fce587f6160, h:-1, <TCP '127.0.0.1 : 8085'>, <TCP '127.0.0.1 : 44828'> FD Closed>, duration: 00:20:04.181045 (hh:mm:ss.us))
--> [context]zKq7AVECAAAAALGqfAALdnB4ZAAAMJwqbGlidm1hY29yZS5zbwAAUC4bAJ6yGACSmRoAFFAlAAU9IwA9CiMABg0jAOHzKgHUcwBsaWJwdGhyZWFkLnNvLjAAAr2LDmxpYmMuc28uNgA=[/context]
2018-08-24T07:38:16.856Z error vpxd[40935] [Originator@6876 sub=SoapAdapter.HTTPService] Failed to read request; stream: <io_obj p:0x00007fce304ab640, h:-1, <TCP '127.0.0.1 : 8085'>, <TCP '127.0.0.1 : 44824'> FD Closed>, error: N7Vmacore16TimeoutExceptionE(Operation timed out: Stream: <io_obj p:0x00007fce304ab640, h:-1, <TCP '127.0.0.1 : 8085'>, <TCP '127.0.0.1 : 44824'> FD Closed>, duration: 00:20:04.037826 (hh:mm:ss.us))
--> [context]zKq7AVECAAAAALGqfAALdnB4ZAAAMJwqbGlidm1hY29yZS5zbwAAUC4bAJ6yGACSmRoAFFAlAAU9IwA9CiMABg0jAOHzKgHUcwBsaWJwdGhyZWFkLnNvLjAAAr2LDmxpYmMuc28uNgA=[/context]
2018-08-24T07:38:16.856Z error vpxd[04025] [Originator@6876 sub=SoapAdapter.HTTPService] Failed to read request; stream: <io_obj p:0x00007fce3806c300, h:-1, <TCP '127.0.0.1 : 8085'>, <TCP '127.0.0.1 : 53558'> FD Closed>, error: N7Vmacore16TimeoutExceptionE(Operation timed out: Stream: <io_obj p:0x00007fce3806c300, h:-1, <TCP '127.0.0.1 : 8085'>, <TCP '127.0.0.1 : 53558'> FD Closed>, duration: 00:00:49.977699 (hh:mm:ss.us))
